Add Course Landing View

ferrarafer opened the pull request:
- Added CourseLandingView
- Refactored code for common widgets use

ferrarafer committed:
Add CourseLandingView

ferrarafer committed:
Refactor code for common widgets use

FilledStacks commented on lib/ui/views/course_landing/course_landing_view.dart:
```diff
@@ -0,0 +1,29 @@
+  CourseLandingViewModel viewModelBuilder(
+    BuildContext context,
+  ) =>
+      CourseLandingViewModel(courseId: 'flutter-web');
```
We actually want the path of this to include this id and get it from there. We should discuss this when you're up but what I'm currently thinking is:

- `course/flutter-web` shows `CourseLandingView`
- `course/flutter-web/chapterId` shows `CourseDetailsView` with the chapter selected. 

This will require some updates, which are luckily now all detailed in the [new navigation video ](https://youtu.be/cGdNRnBdWpw)

ferrarafer replied:
I don't know if there is a bug on the `RouterService` or I'm doing something wrong but when I navigate to `CourseDetailsView` with `chapterId`, the chapterId is not taken into account because is null and the first chapterId is always chosen.

Also, I'm not sure how we are managing the routing here because for what I think, CourseLandingView and CourseDetailsView should be the same. Why the content of CourseLandingView is not on CourseDetailsView?

With the current setup, we must pass CourseLandingView first to the Stacked routes, if not it is ignored because CourseDetailsView route would match first.

Let's talk about this tomorrow. I believe our routes should be like below:

`/courses/:courseId/chapters/:chapterId`

being resources always in plural. If you see any other site, the convention is always like this.

FilledStacks replied:
@ferrarafer 

- `CourseLandingView` is the view that shows the course overview and the course trailer. 
- `CorseDetailsView` is the view where they consume the content
- `CourseChapterView` is nested in the `CourseDetailsView`

we'll discuss later and you can take me through it and why it's not working. I agree on the plural for the path, but I don't think we need the `/chapter` since we're not really navigating to chapters. But we'll discuss soon.

FilledStacks commented on lib/ui/views/course_landing/course_landing_view.tablet.dart:
This can be completely deleted.

ferrarafer replied:
Done ✅

FilledStacks commented on lib/ui/views/course_landing/widgets/course_landing_overview.dart:
```diff
@@ -0,0 +1,71 @@
+  const CourseLandingOverview({Key? key}) : super(key: key);
+
+  static const String courseDetails = '''
+The course teaches patterns, file layouts and development principles used to build the FilledStacks Academy. Yes, the website you’re using right now. 
```
This we should get from `course.description`

ferrarafer replied:
I forgot to remove the hardcoding data 😅 . Done ✅ .

FilledStacks commented on lib/ui/views/course_landing/widgets/course_landing_overview.dart:
```diff
@@ -0,0 +1,71 @@
+              verticalSpaceLarge,
+              CourseBenefitsCard(
+                benefits: [
+                  Benefit(
```
I like this abstraction. Can we please add these into the models so we can have it in the db.

ferrarafer replied:
Done ✅ , refactored from `Benefit` to `Perk` because you called like this today on the call and I like more that name 😄 .

FilledStacks commented on test/viewmodels/course_landing_viewmodel_test.dart:
You're making me look bad with your unit tests! 😂  Which is really awesome. Well done ... I'll ensure to do my tests as well.

ferrarafer replied:
We both know that if the unit tests are not there is because we didn't have time to make them, not because we didn't want to. In this case, I had some minutes before dinner was ready, so I made them 💪 .

FilledStacks reviewed (changes requested):
Happy with the layouts and functionality. Biggest changes are:
- Get the data we're hardcoding from the model (Add it into the fake data)
- Create all properties on the model that's required for the new data
- Correct the path of the new view using the comments below.

ferrarafer committed:
Update stacked_firebase_package

- Updated `stacked_firebase_package`
- Removed `firebase_auth` package dependency because the Firebase User is
  now exposed on `stacked_firebase_package`

ferrarafer committed:
Refactor CourseBenefitsCard to CoursePerksCard

- Refactored CourseBenefitsCard to CoursePerksCard
- Updated Course model adding description and perks properties

ferrarafer committed:
Updated fake data on CourseService

ferrarafer committed:
Update path from course to courses

ferrarafer committed:
Update CourseLandingOverview widget

ferrarafer committed:
Change navigateToCourse route

ferrarafer committed:
Pass courseId to CourseLandingView

ferrarafer committed:
Fix getCourseForId function

When the course was not find, the firstWhere swallowed the StateError
and the state of the app was not clear. I use a try and catch to catch
the StateError and return null.

ferrarafer committed:
Add UnknownView and redirect to it after no match

ferrarafer committed:
Redirect to UnknownView on course not find

ferrarafer committed:
Add pod_player package

ferrarafer committed:
Add VimeoPlayer common widget

ferrarafer committed:
Update CourseLandingTrailer widget

ferrarafer committed:
Fix issue with sync function inside async function

FilledStacks committed:
Merge pull request #6 from FilledStacks/feat/404-error-page

feat: add 404 error view and redirection fallback

ferrarafer committed:
Update pod_player package, using fork

FilledStacks commented:
@ferrarafer has the path updates been applied?

ferrarafer committed:
Merge branch 'tutorials/05-course-landing' into feat/vimeo-player

ferrarafer commented:
> @ferrarafer has the path updates been applied?

I fixed it inside https://github.com/FilledStacks/academy/tree/feat/vimeo-player branch.

```
Future<Course?> getCourseForId(String id) async {
    try {
      // Delay added to avoid execute a sync function as Future
      await Future.delayed(const Duration(microseconds: 100));
      return courses.firstWhere((element) => element.id == id);
    } on StateError catch (_) {
      log.w('Course with id equal "$id" not found');
      return null;
    } catch (e) {
      log.e(e.toString());
      return null;
    }
  }
```

FilledStacks committed:
Merge pull request #7 from FilledStacks/feat/vimeo-player

feat: add Vimeo support
